pkg/admin/prerun/prerun.go

```diff
@@ -109,6 +115,109 @@ func (pr *PreRun) Perform() error {
+func (pr *PreRun) checkPreconditions() (bool, error) {
+	dataExists, err := util.PathExists(config.DataDir)
+	if err != nil {
+		return false, fmt.Errorf("failed to check if data directory already exists: %w", err)
+	}
+
+	versionExists, err := util.PathExists(versionFilePath)
+	if err != nil {
+		return false, fmt.Errorf("checking if version metadata exists failed: %w", err)
+	}
+
+	healthExists, err := util.PathExists(healthFilepath)
+	if err != nil {
+		return false, fmt.Errorf("failed to check if health file already exists: %w", err)
+	}
+
+	klog.InfoS("Existence of important paths",
+		"data-exists?", dataExists,
+		"version-exists?", versionExists,
+		"health-exists?", healthExists,
+	)
+
+	/*
+		| id  | data | version | health | comment                                                                                                     |
+		| --- | ---- | ------- | ------ | ----------------------------------------------------------------------------------------------------------- |
+		| 1   | 0    | 0       | 0      | first, clean start of MicroShift                                                                            |
+		| 2   | 0    | 0       | 1      | data removed manually, but health/backups kept                                                              |
+		| 3   | 0    | 1       | 0      | impossible to detect right now [0]                                                                          |
+		| 4   | 0    | 1       | 1      | impossible to detect right now [0]                                                                          |
+		| 5   | 1    | 0       | 0      | upgrade from 4.13                                                                                           |
+		| 6   | 1    | 0       | 1      | first boot failed very early, or upgrade from 4.13 failed (e.g. healthcheck didn't see service being ready) |
+		| 7   | 1    | 1       | 0      | first start, rebooted before green/red scripts managed to write health info                                 |
+		| 8   | 1    | 1       | 1      | regular                                                                                                     |
+
+		[0] it would need a comprehensive check if data exists, not just existence of /var/lib/microshift
+	*/
+
+	if !dataExists {
+		// 1
+		if !healthExists {
+			klog.InfoS("Neither data dir nor health file exist - assuming first start of MicroShift")
+			return false, nil
+		}
+
+		// 2
+		// TODO: Health needs to be extended into a history of boots and their health
+		// so prerun can look into the past and decide if a backup should be restored
+		// for currently running deployment
+		return false, fmt.Errorf("TODO: data is missing, but health file exists")
+	}
+
+	// 5
+	if !versionExists && !healthExists {
+		klog.InfoS("Data dir exists, but health and version files are missing: assuming upgrade from 4.13")
+		return false, pr.upgradeFrom413()
```

Comment:
Why not perform the pre-run process when moving from 4.13?

Reply:
I tried that in first version (https://github.com/openshift/microshift/commit/333216d4bab887a7f4aacb2ff2b866f89b1c2acd#diff-e78e17834c76663b756393d04ebe7aceb8ea4059d44b7d2ae3ee4e9999aead6eR65) but I didn't like the workaround to fit in that process (manually constructing "4.13" + "healthy" and generating backup name from deployId without boot id)

Reply:
I also expect that when we get to solving TODOs in this function, it might get renamed to resolve these special cases so all of this will be part of pre-run and what we have today will be "regular flow"